server/integration/endpoint/src/main/java/org/infinispan/server/endpoint/subsystem/MultiTenantRouterService.java

```diff
@@ -0,0 +1,127 @@
+import org.jboss.msc.service.StartException;
+import org.jboss.msc.service.StopContext;
+import org.jboss.msc.value.InjectedValue;
+
+import java.net.InetSocketAddress;
+import java.util.HashMap;
+import java.util.Optional;
+
+import static org.infinispan.server.endpoint.EndpointLogger.ROOT_LOGGER;
+
+/**
+ * Multi tenant router service
+ *
+ * @author Sebastian Łaskawiec
+ */
+class MultiTenantRouterService implements Service<MultiTenantRouter> {
+
+    static class HotRodRouting {
+        private final InjectedValue<HotRodServer> hotRod = new InjectedValue<>();
+        private final InjectedValue<SecurityRealm> securityRealm = new InjectedValue<>();
+
+        public InjectedValue<HotRodServer> getHotRod() {
+            return hotRod;
+        }
+
+        public InjectedValue<SecurityRealm> getSecurityRealm() {
+            return securityRealm;
+        }
+    }
+
+    private static final String DEFAULT_NAME = "Multitenant Router";
+
+    private final InjectedValue<SocketBinding> socketBinding = new InjectedValue<>();
+    private final InjectedValue<HotRodServer> hotrod = new InjectedValue<>();
+    private final java.util.Map<String, HotRodRouting> sniRouting = new HashMap<>();
+
+    private final String name;
+    private final MultiTenantRouterConfigurationBuilder configurationBuilder;
+    private Optional<MultiTenantRouter> router = Optional.empty();
+
+    MultiTenantRouterService(MultiTenantRouterConfigurationBuilder configurationBuilder, Optional<String> serverName) {
+        this.name = constructServerName(serverName);
+        this.configurationBuilder = configurationBuilder;
+    }
+
+    private final String constructServerName(Optional<String> name) {
+        return name.orElse(DEFAULT_NAME);
+    }
+
+    @Override
+    public synchronized void start(final StartContext context) throws StartException {
+        ROOT_LOGGER.endpointStarting(name);
+        try {
+            SocketBinding socketBinding = getSocketBinding().getValue();
+            InetSocketAddress socketAddress = socketBinding.getSocketAddress();
+            configurationBuilder.ip(socketAddress.getAddress());
+            configurationBuilder.port(socketAddress.getPort());
+
+            sniRouting.entrySet().stream().forEach(e -> {
+                HotRodServer hotRod = e.getValue().getHotRod().getValue();
+                SecurityRealm securityRealm = e.getValue().getSecurityRealm().getValue();
+                String sniHostName = e.getKey();
+
+                SniNettyRouteSource source = new SniNettyRouteSource(sniHostName, securityRealm.getSSLContext());
+                NettyHandlerRouteDestination destination = new NettyHandlerRouteDestination(hotRod.getQualifiedName(), hotRod.getInitializer());
+
+                configurationBuilder.routing().add(new Route<>(source, destination));
+            });
+
+            router = Optional.of(new MultiTenantRouter(configurationBuilder.build()));
+            router.ifPresent(r -> r.start());
```

Comment:
This will always be true, maybe just hold a reference to the `MultiTenantRouter` instance and call start directly.